### ISSUE TITLE
Shadowing imported names results in TC004 false-positives

### ISSUE THREAD

**Issue by shiftinv** [snok/flake8-type-checking#94]
First of all, thank you for this neat plugin! It has been helping out quite a bit with managing type-checking imports :p

If a name that was imported in a type-checking block is shadowed by a variable or parameter of the same name, a TC004 error is emitted, asking to move the import out of the type-checking block:

```py
from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from typing import List as list_type  # this would be used in other places, left out for brevity

def example():
    list_type = list  # shadowing imported name
    return list_type([42])
```
```sh
$ flake8 --select TC test.py
test.py:7:1: TC004 Move import 'list_type' out of type-checking block. Import is used for more than type hinting.
```

This is obviously a crafted example, a more practical one can be seen here - the `emoji` module is imported from `.types` in a type-checking block to later be used in annotations:
https://github.com/DisnakeDev/disnake/blob/297713205d20c90086c29d168b279ff7ac2cc5a7/disnake/http.py#L76-L82
https://github.com/DisnakeDev/disnake/blob/297713205d20c90086c29d168b279ff7ac2cc5a7/disnake/http.py#L1613

Further down, a method also takes a parameter named `emoji`, shadowing the imported module; this appears to be considered runtime usage of the import instead:
https://github.com/DisnakeDev/disnake/blob/297713205d20c90086c29d168b279ff7ac2cc5a7/disnake/http.py#L710-L712

Naturally, shadowing global variables/imports/... is generally discouraged, though I'd argue it's not an issue in this scenario as the shadowed variable/import is only used in annotations.

I'm unsure if the ast module allows differentiating between these names, though it probably does as flake8 detects the import in the example above as unused. If this cannot be fixed easily it's no big deal, as just renaming the import would be the easy way out 🤔

**Comment by sondrelg** [comment]
Thanks for opening the issue! I'm away until late next week and will take a proper look at this then 😊 In the meantime, if you're interested in contributing - please feel free to open a PR 👍

**Comment by sondrelg** [comment]
This is a bit of a tricky one. After looking over the issue, it seems like we would need to try to figure out which scope the shadowed variable is contained in. For example, if the shadowed variable is defined like this:

```python
0 from typing import List as l  
1
2 def some_function():
3    x: l[int]  # <-- we use the type import here
4    l = set  # <-- we shadow the import name
5    y = l([1,2,])  # <-- we use the new variable here 
6 
7 def some_other_function(): ...
```

Then we would need to recognize that any use of the node `id` before line 4 should be classified as usage of the import, but usage after line 4, *until* line 5, is not. 

Unfortunately, each of the scopes:

- Global
- (Async/sync) function
- Class 
- (Async/sync) method

Would require completely separate logic, and the plugin doesn't really have a way to determine what scope we're in when handling an `ast.Assign` node. We don't really get the outer context, and I'm not sure if there's a great way to achieve this without a substantial rewrite of the plugin. 

So tl;dr this seems hard to implement at first glance. Do you have any ideas for how it might be implemented?

**Comment by shiftinv** [comment]
Yeah, it does look quite difficult to handle, since variable scoping doesn't seem trivial with the builtin ast module.
Keeping track of scopes manually doesn't sound like a good time either, especially given some of the funky things you can do in Python with nested functions, lambdas, and the like. Not to mention that doing that would likely not be worth it given that this is a fairly obscure edge case, and easy to workaround by just using a different name for the import.

I've been planning to look into how flake8 itself handles it, as it certainly seems possible, but haven't found the time for it yet :/

**Comment by sondrelg** [comment]
I'm glad you see the problem, and I would be very happy to accept a fix for this that *didn't* involve manual mapping of scopes 🙃 If you figure something out, let me know, or feel free to open a PR!

**Comment by sondrelg** [comment]
I'll close this for now, since it's not something we're likely to pursue in the short term. Happy to review code or discuss this more if new ideas surface 🙂
